infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -288,55 +288,46 @@ private void prefixTrim(CorfuPayloadMsg<TrimRequest> msg, ChannelHandlerContext
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_CORRUPTION.payloadMsg(address));
-        }
+  @ServerHandler(type = CorfuMsgType.READ_REQUEST)
+  public void read(CorfuPayloadMsg<ReadRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
+    boolean cacheable = msg.getPayload().isCacheReadResult();
+    if (log.isTraceEnabled()) {
+      log.trace("read: {}, cacheable: {}", msg.getPayload().getAddresses(), cacheable);
     }
 
-    @ServerHandler(type = CorfuMsgType.MULTIPLE_READ_REQUEST)
-    public void multiRead(CorfuPayloadMsg<MultipleReadRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
-        boolean cacheable = msg.getPayload().isCacheReadResult();
-        log.trace("multiRead: {}, cacheable: {}", msg.getPayload().getAddresses(), cacheable);
-
-        ReadResponse rr = new ReadResponse();
-        try {
-            for (Long address : msg.getPayload().getAddresses()) {
-                ILogData logData = dataCache.get(address, cacheable);
-                if (logData == null) {
-                    rr.put(address, LogData.getEmpty(address));
-                } else {
-                    rr.put(address, (LogData) logData);
-                }
-            }
-            r.sendResponse(ctx, msg, CorfuMsgType.READ_RESPONSE.payloadMsg(rr));
-        } catch (DataCorruptionException e) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_CORRUPTION.msg());
+    ReadResponse rr = new ReadResponse();
+    long currentRead = Address.NON_ADDRESS;
+    try {
+      for (Long address : msg.getPayload().getAddresses()) {
+        currentRead = address;
+        ILogData logData = dataCache.get(address, cacheable);
+        if (logData == null) {
+          rr.put(address, LogData.getEmpty(address));
+        } else {
+          rr.put(address, (LogData) logData);
         }
+      }
+      r.sendResponse(ctx, msg, CorfuMsgType.READ_RESPONSE.payloadMsg(rr));
+    } catch (DataCorruptionException e) {
+      log.error(
+          "Data corruption exception while reading addresses {}",
+          msg.getPayload().getAddresses(),
+          e);
+      r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_CORRUPTION.payloadMsg(currentRead));
     }
+  }
 
     @ServerHandler(type = CorfuMsgType.INSPECT_ADDRESSES_REQUEST)
     public void inspectAddresses(CorfuPayloadMsg<InspectAddressesRequest> msg,
                                  ChannelHandlerContext ctx, IServerRouter r) {
-        log.trace("inspectAddresses: {}", msg.getPayload().getAddresses());
+        List<Long> addresses = msg.getPayload().getAddresses();
+        log.trace("inspectAddresses: {}", addresses);
         InspectAddressesResponse inspectResponse = new InspectAddressesResponse();
+
+    long currentAddress = Address.NON_ADDRESS;
         try {
-            for (long address : msg.getPayload().getAddresses()) {
+            for (long address : addresses) {
+                currentAddress = address;
```

Comment:
`address` is primitive long, do we really need `currentAddress`?

Reply:
Done.